rmf_site_editor/src/sdf_loader.rs

```diff
@@ -60,25 +68,228 @@ impl AssetLoader for SdfLoader {
+                    return Err(SdfError::UnsupportedAssetSource(format!(
+                        "Unable to extract organization name from asset source [{}]",
+                        uri
+                    )));
+                }
+            }
+            AssetSource::Local(ref mut p) | AssetSource::Package(ref mut p) => {
+                // Search for a model with the requested name in the same folder as the sdf file
+                // Note that this will not play well if the requested model shares files with other
+                // models that are placed in different folders or are in fuel, but should work for
+                // most local, self contained, models.
+                *p = if let Some(model_folder) = p.rsplitn(3, "/").skip(2).next() {
+                    model_folder.to_owned() + "/" + stripped
+                } else {
+                    return Err(SdfError::UnsupportedAssetSource(format!(
+                        "Unable to extract model folder from asset source [{}]",
+                        uri
+                    )));
+                }
+            }
+        }
+        Ok(asset_source)
+    } else if let Some(path_idx) = path.rfind("/") {
+        // It's a path relative to this model, remove file and append uri
+        let (sdf_path, _sdf_file_name) = path.split_at(path_idx);
+        AssetSource::try_from((sdf_path.to_owned() + "/" + uri).as_str())
+            .map_err(SdfError::UnsupportedAssetSource)
+    } else {
+        Err(SdfError::UnsupportedAssetSource(format!(
+            "Invalid SDF model path, Path is [{}] and model uri is [{}]",
+            path, uri
+        )))
+    }?;
+    Ok(asset_source)
+}
+
+fn parse_scale(scale: &Option<Vector3d>) -> Scale {
+    match scale {
+        Some(v) => Scale(Vec3::new(v.0.x as f32, v.0.y as f32, v.0.z as f32)),
+        None => Scale::default(),
+    }
+}
+
+fn parse_pose(pose: &Option<SdfPose>) -> Pose {
+    if let Some(pose) = pose.clone().and_then(|p| p.get_pose().ok()) {
+        let rot = pose.rotation.euler_angles();
+        Pose {
+            trans: [
+                pose.translation.x as f32,
+                pose.translation.y as f32,
+                pose.translation.z as f32,
+            ],
+            rot: Rotation::EulerExtrinsicXYZ([
+                Angle::Rad(rot.0 as f32),
+                Angle::Rad(rot.1 as f32),
+                Angle::Rad(rot.2 as f32),
+            ]),
+        }
+    } else {
+        Pose::default()
+    }
+}
+
+fn spawn_geometry<'a, 'b>(
+    world: &'a mut World,
+    geometry: &'a SdfGeometry,
+    visual_name: &'a str,
+    pose: &'a Option<SdfPose>,
+    load_context: &'a mut LoadContext<'b>,
+    is_static: bool,
+) -> Result<Option<Entity>, SdfError> {
+    let pose = parse_pose(pose);
+    let geometry = match geometry {
+        SdfGeometry::Mesh(mesh) => Some(
+            world
+                .spawn(Model {
+                    name: NameInSite(visual_name.to_owned()),
```

Comment:
Fully agree, `Model` served well in the early days of the codebase when each model was a single GLB file but since we started to parse SDFs with hierarchies, include primitive shapes, split of collision and visual meshes, it became very complex.
Ideally the perfect design would allow granularity to encapsulate all the components of a `Model` into a top level object (for example to know where to fetch a model if it comes from fuel) but at the same time would allow inspecting its parts (toggling visual / collision meshes, plus some other features needed in workcells that work with individual submeshes)